### PR TITLE
Skip GPU driver install with golden image

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: Completely skip install of GPU drivers if pre-installed on golden image. Add specific exit codes for systemctlEnableAndStart.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**If applicable**:
- [ ] documentation
- [ ] unit tests
- [ ] tested backward compatibility (ie. deploy with previous version, upgrade with this branch)

**Release note**:

```release-note
```

### CODE REVIEW

File: parts/k8s/kubernetesconfigs.sh [hunk 1]
```diff
@@ -257,19 +257,19 @@ function ensureKubelet() {
     wait_for_file 1200 1 $KUBECONFIG_FILE || exit $ERR_FILE_WATCH_TIMEOUT
     KUBELET_RUNTIME_CONFIG_SCRIPT_FILE=/opt/azure/containers/kubelet.sh
     wait_for_file 1200 1 $KUBELET_RUNTIME_CONFIG_SCRIPT_FILE || exit $ERR_FILE_WATCH_TIMEOUT
-    systemctlEnableAndStart kubelet
+    systemctlEnableAndStart kubelet || exit $ERR_KUBELET_START_FAIL
     # Don't start the kubelet-health-monitor systemd unit (yet).
     # KUBELET_HEALTH_MONITOR_SYSTEMD_FILE=/etc/systemd/system/kubelet-health-monitor.service
     # wait_for_file 1200 1 $KUBELET_HEALTH_MONITOR_SYSTEMD_FILE || exit $ERR_FILE_WATCH_TIMEOUT
-    # systemctlEnableAndStart kubelet-health-monitor
+    # systemctlEnableAndStart kubelet-health-monitor || exit $ERR_SYSTEMCTL_START_FAIL
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesconfigs.sh]: @mboersma FYI this is going to conflict with you PR

File: parts/k8s/kubernetesconfigs.sh [hunk 2]
```diff
@@ -392,7 +392,7 @@ configAddons() {
 }
 
 configGPUDrivers() {
-    sh nvidia-drivers-$GPU_DV --silent --accept-license --no-drm --dkms --utility-prefix="${GPU_DEST}" --opengl-prefix="${GPU_DEST}"
+    sh /usr/local/nvidia/nvidia-drivers-$GPU_DV --silent --accept-license --no-drm --dkms --utility-prefix="${GPU_DEST}" --opengl-prefix="${GPU_DEST}" || exit $ERR_GPU_DRIVERS_START_FAIL
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigs.sh]: Can we also add the full path to the curl output destination in `parts/k8s/kubernetesinstalls.sh`?

File: parts/k8s/kubernetesinstalls.sh [hunk 3]
```diff
@@ -30,35 +30,30 @@ function installDeps() {
...
+    apt_get_update
+    retrycmd_if_failure 30 5 300 apt-get install -y linux-headers-$(uname -r) gcc make dkms || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
+    retrycmd_if_failure 30 5 300 apt-get -o Dpkg::Options::="--force-confold" install -y nvidia-docker2=$NVIDIA_DOCKER_VERSION+docker$DOCKER_VERSION nvidia-container-runtime=$NVIDIA_CONTAINER_RUNTIME_VERSION+docker$DOCKER_VERSION || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
+    pkill -SIGHUP dockerd
+    mkdir -p $GPU_DEST
+    cd $GPU_DEST
+    # Download the .run file from NVIDIA.
+    # Nvidia libraries are always install in /usr/lib/x86_64-linux-gnu, and there is no option in the run file to change this.
+    # Instead we use Overlayfs to move the newly installed libraries under /usr/local/nvidia/lib64
+    retrycmd_if_failure 5 10 60 curl -fLS https://us.download.nvidia.com/tesla/$GPU_DV/NVIDIA-Linux-x86_64-$GPU_DV.run -o nvidia-drivers-$GPU_DV || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
```
Comment (jackfrancis) [inline parts/k8s/kubernetesinstalls.sh]: oh. yeah, I mean here

File: parts/k8s/kubernetesconfigs.sh [hunk 4]
```diff
@@ -392,7 +392,7 @@ configAddons() {
 }
 
 configGPUDrivers() {
-    sh nvidia-drivers-$GPU_DV --silent --accept-license --no-drm --dkms --utility-prefix="${GPU_DEST}" --opengl-prefix="${GPU_DEST}"
+    sh $GPU_DEST/nvidia-drivers-$GPU_DV --silent --accept-license --no-drm --dkms --utility-prefix="${GPU_DEST}" --opengl-prefix="${GPU_DEST}" || exit $ERR_GPU_DRIVERS_START_FAIL
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesconfigs.sh]: @jackfrancis turns out there was a var for it

File: parts/k8s/kubernetesinstalls.sh [hunk 5]
```diff
@@ -30,35 +30,27 @@ function installDeps() {
...
+    cat /tmp/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
+    apt_get_update
+    retrycmd_if_failure 30 5 300 apt-get install -y linux-headers-$(uname -r) gcc make dkms || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
+    retrycmd_if_failure 30 5 300 apt-get -o Dpkg::Options::="--force-confold" install -y nvidia-docker2=$NVIDIA_DOCKER_VERSION+docker$DOCKER_VERSION nvidia-container-runtime=$NVIDIA_CONTAINER_RUNTIME_VERSION+docker$DOCKER_VERSION || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
+    pkill -SIGHUP dockerd
+    # Download the .run file from NVIDIA.
+    # Nvidia libraries are always install in /usr/lib/x86_64-linux-gnu, and there is no option in the run file to change this.
+    # Instead we use Overlayfs to move the newly installed libraries under /usr/local/nvidia/lib64
+    retrycmd_if_failure 5 10 60 curl -fLS https://us.download.nvidia.com/tesla/$GPU_DV/NVIDIA-Linux-x86_64-$GPU_DV.run -o $GPU_DEST/nvidia-drivers-$GPU_DV || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesinstalls.sh]: updated output with absolute path and removed `cd`

File: parts/k8s/kubernetesinstalls.sh [hunk 6]
```diff
@@ -30,35 +30,27 @@ function installDeps() {
...
-        # Installing nvidia-docker, setting nvidia runtime as default and restarting docker daemon
-        retrycmd_if_failure_no_stats 180 1 5 curl -fsSL https://nvidia.github.io/nvidia-docker/gpgkey > /tmp/aptnvidia.gpg || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
-        cat /tmp/aptnvidia.gpg | apt-key add -
-        retrycmd_if_failure_no_stats 180 1 5 curl -fsSL https://nvidia.github.io/nvidia-docker/ubuntu16.04/amd64/nvidia-docker.list > /tmp/nvidia-docker.list || exit  $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
-        cat /tmp/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list
-        apt_get_update
-        retrycmd_if_failure 30 5 300 apt-get install -y linux-headers-$(uname -r) gcc make dkms || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
-        retrycmd_if_failure 30 5 300 apt-get -o Dpkg::Options::="--force-confold" install -y nvidia-docker2=$NVIDIA_DOCKER_VERSION+docker$DOCKER_VERSION nvidia-container-runtime=$NVIDIA_CONTAINER_RUNTIME_VERSION+docker$DOCKER_VERSION || exit $ERR_GPU_DRIVERS_INSTALL_TIMEOUT
-        pkill -SIGHUP dockerd
-        mkdir -p $GPU_DEST
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesinstalls.sh]: removed this second mkdir, already done above

**Pull request merged**